File: assembler folder/assembler.py
from parser import Parser
from binary_code import BinaryCode
from symbol_table import SymbolTable

def assembler(input_file, output_file):
    #First pass
    parser = Parser(input_file)
    symbol_table = SymbolTable()
    instruction_counter = 0

    while parser.has_more_commands():
        parser.advance()
        command_type = parser.command_type()
        
        if command_type == 'L_COMMAND':
            symbol = parser.symbol()
            symbol_table.add_entry(symbol, instruction_counter)
        else:
            instruction_counter += 1

    # Second pass
    parser = Parser(input_file)
    code = BinaryCode()
    binary_instructions = []

    while parser.has_more_commands():
        parser.advance()
        command_type = parser.command_type()

        if parser.command_type() == 'A_COMMAND':
            symbol = parser.symbol()
            if symbol.isdigit():
                value = int(symbol)
            else:
                value = symbol_table.add_variable(symbol)
            binary = format(value, '016b')
            binary_instructions.append(binary)
            
        elif parser.command_type() == 'C_COMMAND':
            comp = code.comp(parser.comp())
            dest = code.dest(parser.dest())
            jump = code.jump(parser.jump())
            binary = f'111{comp}{dest}{jump}'
            binary_instructions.append(binary)

    with open(output_file, 'w', newline='') as f:
        for i, instruction in enumerate(binary_instructions):
            if i < len(binary_instructions) - 1:
                f.write(f"{instruction}\n")
            else:
                f.write(instruction)

if __name__ == '__main__':
    input_file = 'rect/Rect.asm'
    output_file = 'Rect1.hack'
    assembler(input_file, output_file)

File: assembler folder/Assembler.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class Assembler {
    public static void assemble(String inputFile, String outputFile) throws IOException {
        // First pass
        Parser parser = new Parser(inputFile);
        SymbolTable symbolTable = new SymbolTable();
        int instructionCounter = 0;

        while (parser.hasMoreCommands()) {
            parser.advance();
            String commandType = parser.commandType();

            if (commandType.equals("L_COMMAND")) {
                String symbol = parser.symbol();
                symbolTable.addEntry(symbol, instructionCounter);
            } else {
                instructionCounter++;
            }
        }

        // Second pass
        parser = new Parser(inputFile);
        BinaryCode code = new BinaryCode();
        List<String> binaryInstructions = new ArrayList<>();

        while (parser.hasMoreCommands()) {
            parser.advance();

            if (parser.commandType().equals("A_COMMAND")) {
                String symbol = parser.symbol();
                int value;
                if (symbol.matches("\\d+")) {
                    value = Integer.parseInt(symbol);
                } else {
                    value = symbolTable.addVariable(symbol);
                }
                String binary = String.format("%16s", Integer.toBinaryString(value))
                                    .replace(' ', '0');
                binaryInstructions.add(binary);

            } else if (parser.commandType().equals("C_COMMAND")) {
                String comp = code.comp(parser.comp());
                String dest = code.dest(parser.dest());
                String jump = code.jump(parser.jump());
                String binary = "111" + comp + dest + jump;
                binaryInstructions.add(binary);
            }
        }

        // Write output file
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile))) {
            for (int i = 0; i < binaryInstructions.size(); i++) {
                writer.write(binaryInstructions.get(i));
                if (i < binaryInstructions.size() - 1) {
                    writer.write("\n");
                }
            }
        }
    }

    public static void main(String[] args) {
        try {
            String inputFile = "pong/Pong.asm";
            String outputFile = "Pong2.hack";
            assemble(inputFile, outputFile);
        } catch (IOException e) {
            System.err.println("Error processing file: " + e.getMessage());
            System.exit(1);
        }
    }
}
